src/language/pattern-types.md

```diff
@@ -0,0 +1,333 @@
+var (i as int, s as String) = record;
+```
+
+Cast patterns will [throw][] if the value doesn't have the stated type.
+Like the [null-assert pattern](#null-assert), this lets you forcibly assert the
+expected type of some destructured value.
+
+## Constant	
+
+`123, null, 'string', math.pi, SomeClass.constant, const Thing(1, 2), const (1 + 2)`
+
+Constant patterns allow simple literals and references to named constants to be
+used directly as patterns. They match when the value is equal to the constant.
+
+Some valid constant expressions may not be valid constant patterns. In that case,
+to distinguish a constant expression from a constant pattern, prefix the expression
+with `const` and wrap it in parenthesis so it is unambiguosly an expression and
+not a constant pattern:
+
+```dart
+// List or map pattern:
+case [a, b]: ...
+
+// List or map literal:
+case const [a, b]: ...
+```
+
+## Identifier	
+
+`foo, _`
+
+Identifier patterns may behave like a [constant pattern](#constant) or like a
+[variable pattern](#variable), depending on the context where it appears:
+
+- [Declaration][] context: declares a new variable with identifier name:
+  `var (a, b) = (1, 2);`
+- [Assignment][] context: assigns to existing variable with identifier name:
+  `(a, b) = (3, 4);`
+- [Matching][] context: treated as a named constant pattern (unless its name is `_`):
+  ```dart
+  const c = 1;
+  switch (2) {
+    case c: print('match $c');
+    default: print('no match');    // Prints "no match".
+   }
+  ``` 
+- [Wildcard](#wildcard) identifier in any context: matches any value and discards it:
+  `case [_, var y, _]: print('The middle element is $y');`
+
+## List
+
+`[subpattern1, subpattern2]`
+
+List patterns match values that implement [`List`][], and then recursively match
+its subpatterns against the list's elements to destructure them by position.
+
+List patterns require that the pattern match the entire list, but you can use
+[rest elements](#rest-element) as a place holder.
+
+### Rest element
+
+List patterns can contain _one_ rest element (`...`) which allows matching lists
+of arbitrary lengths.
+
+```dart
+var [a, b, ..., c, d] = [1, 2, 3, 4, 5, 6, 7];
+print('$a $b $c $d'); // Prints "1 2 6 7".
+```
+
+Rest elements can also have subpatterns that collects elements that don't match
+the other subpatterns in the list, into a new list:
+
+```dart
+var [a, b, ...rest, c, d] = [1, 2, 3, 4, 5, 6, 7];
+print('$a $b $rest $c $d'); // Prints "1 2 [3, 4, 5] 6 7".
+```
+
+## Logical-and	
+
+`subpattern1 && subpattern2`
+
+A pair of patterns separated by `&&` matches only if both subpatterns match. If the
+left branch does not match, the right branch is not evaluated. 
+
+## Logical-or
+
+`subpattern1 || subpattern2`
+
+A pair of patterns separated by `||` matches if either of the branches match.
+If the left branch matches, the right branch is not evaluated. 
+
+Logical-or patterns can be nested inside a destructuring pattern. 
```

Comment:
Actually any pattern can be nested inside a destructuring pattern. You can go wild: `case Foo([a]! && {1: b}? || Bar() as Baz)`. So I'd probably not call this out here since it suggests, at least, that `&&` *can't* be nested.

Reply:
I can add the line "Any [kind of pattern][] can be nested inside a destructuring pattern" to the Destructuring section on the main page.